ntoskrnl/lpc/complete.c

```diff
@@ -315,11 +318,83 @@ NtAcceptConnectPort(OUT PHANDLE PortHandle,
+                                           MmSectionObjectType,
+                                           PreviousMode,
+                                           (PVOID *)&SectionToMap,
+                                           NULL);
+
+        if (NT_SUCCESS(Status))
+        {
+            Status = MmMapViewOfSection(SectionToMap,
+                                        PsGetCurrentProcess(),
+                                        &ServerPort->ServerSectionBase,
+                                        0,
+                                        0,
+                                        &SectionOffset,
+                                        &CapturedServerView.ViewSize,
+                                        ViewUnmap,
+                                        0,
+                                        PAGE_READWRITE);
+
+            if (NT_SUCCESS(Status))
+            {
+                /* Section mapped into server process */
+                if (ServerPort->MappingProcess == NULL)
+                {
+                    ServerPort->MappingProcess = PsGetCurrentProcess();
+                    ObReferenceObject(ServerPort->MappingProcess);
+                }
+                CapturedServerView.SectionOffset = SectionOffset.QuadPart;
+                CapturedServerView.ViewBase = ServerPort->ServerSectionBase;
+
+                SectionOffset.QuadPart = CapturedServerView.SectionOffset;
+
+                ViewSize = CapturedServerView.ViewSize;
+                Status = MmMapViewOfSection(SectionToMap,
+                                            ClientProcess,
+                                            &ClientPort->ServerSectionBase,
+                                            0,
+                                            0,
+                                            &SectionOffset,
+                                            &ViewSize,
+                                            ViewUnmap,
+                                            0,
+                                            PAGE_READWRITE);
+                if (NT_SUCCESS(Status))
+                {
+                    /* Section mapped to client process */
+                    if (ClientPort->MappingProcess == NULL)
+                    {
+                        ClientPort->MappingProcess = ClientProcess;
+                        ObReferenceObject(ClientProcess);
+                    }
+                    CapturedServerView.ViewRemoteBase = ClientPort->ServerSectionBase;
+                    ConnectMessage->ServerView.ViewBase = ClientPort->ServerSectionBase;
+                    ConnectMessage->ServerView.ViewSize = ViewSize;
+                }
+                else
+                {
+                    ObDereferenceObject(ServerPort);
+                }
+            }
+            else
+            {
+                ObDereferenceObject(ServerPort);
+            }
+            ObDereferenceObject(SectionToMap);
+        }
+        else
+        {
```

Comment:
I think that all these repetitive `ObDereferenceObject(ServerPort);` could be written just once, with a separate block:
```c
if (!NT_SUCCESS(Status))
    ObDereferenceObject(ServerPort);
```
at the end of the whole code of this `if (NT_SUCCESS(Status) && ServerView)` big block.
(Since all the code inside it is a nesting of "do_a_call; if (NT_SUCCESS(Status)) {do_something_else;}".)

Reply:
Yes, probably. Feel free to refactor this.